arcgis-ios-sdk-samples/Edit data/Edit and sync features/EditAndSyncFeaturesViewController.swift

```diff
@@ -0,0 +1,280 @@
+                        // For each layer in the serice, add a layer to the map.
+                        let layerURL = self.featureServiceURL?.appendingPathComponent(String(index))
+                        let featureTable = AGSServiceFeatureTable(url: layerURL!)
+                        let featureLayer = AGSFeatureLayer(featureTable: featureTable)
+                        featureLayer.name = layerInfo.name
+                        map.operationalLayers.add(featureLayer)
+                    }
+                }
+            }
+        }
+    }
+    
+    // Clears selection in all layers of the map.
+    private func clearSelection() {
+        for layer in mapView.map!.operationalLayers {
+            if let layer = layer as? AGSFeatureLayer {
+                layer.clearSelection()
+            }
+        }
+    }
+    
+    private func resetUI() {
+        selectedFeature = nil
+        barButtonItem.isEnabled = true
+        barButtonItem.title = syncGeodatabaseTitle
+        instructionsLabel.text = "Tap the sync button"
+    }
+    
+    func geodatabaseDidLoad() {
+        if let error = geodatabase.loadError {
+            self.presentAlert(error: error)
+        } else {
+            // Iterate through the feature tables in the geodatabase and add new layers to the map.
+            self.mapView.map?.operationalLayers.removeAllObjects()
+            for geodatabaseFeatureTable in self.geodatabase.geodatabaseFeatureTables {
+                geodatabaseFeatureTable.load { [weak self, unowned geodatabaseFeatureTable] (error: Error?) in
+                    if let error = error {
+                        self?.presentAlert(error: error)
+                    } else {
+                        // Create a new feature layer from the table and add it to the map.
+                        let featureLayer = AGSFeatureLayer(featureTable: geodatabaseFeatureTable)
+                        self?.mapView.map?.operationalLayers.add(featureLayer)
+                    }
+                }
+            }
+            self.generateJob = nil
+            self.barButtonItem.isEnabled = false
+            self.instructionsLabel.text = "Tap on a feature"
+            self.mapView.touchDelegate = self
+        }
+    }
+    
+    func geodatabaseDidSync() {
+        self.presentAlert(title: "Geodatabase sync sucessful")
+        self.barButtonItem.isEnabled = false
+        self.instructionsLabel.text = "Tap on a feature"
+    }
+    
+    func generateGeodatabase() {
+        // Hide the unnecessary items.
+        barButtonItem.isEnabled = false
+        extentView.isHidden = true
+        
+        // Get the area outlined by the extent view.
+        areaOfInterest = self.extentViewFrameToEnvelope()
+        
+        geodatabaseSyncTask.defaultGenerateGeodatabaseParameters(withExtent: areaOfInterest) { [weak self] (params: AGSGenerateGeodatabaseParameters?, error: Error?) in
+            if let params = params,
+                let self = self {
```

Comment:
I believe I did that to unwrap `params`. Which method (`guard` and unwrapping params twice with `?` or leaving as is unwrapping once at the top) is preferred?

Reply:
In a case like this, I would recommend this:

``` Swift
if let error = error {
  // handle the error in the normal way
}

guard let self = self else { return }

guard let params = params else { return }
```

That way the bulk of the method is not indented too far. This helps with readability too as you're using `guard` to explicitly explain what's important for the sample to continue. And this `guard` statement will `unwrap` params into a non-optional, which is a good thing.

As @philium mentioned elsewhere, if you're just using an optional with inline chaining like `self?...` for calling a function or two, it's ok, but in general for anything longer, like this, (or if you'll be passing that optional in to other functions which don't accept an optional) then you should use an `if let` or `guard let` to unwrap them. I prefer a `guard let` in a case like this because of the reasons given above.